extmod/network_nina.c

```diff
@@ -0,0 +1,591 @@
+    }
+
+    if (self->itf == MOD_NETWORK_STA_IF) {
+        // Initialize WiFi in Station mode.
+        if (nina_connect(ssid, security, key, 0) != 0) {
+            mp_raise_msg_varg(&mp_type_OSError,
+                MP_ERROR_TEXT("could not connect to ssid=%s, sec=%d, key=%s\n"), ssid, security, key);
+        }
+    } else {
+        mp_uint_t channel = args[3].u_int;
+
+        if (security != NINA_SEC_OPEN && security != NINA_SEC_WEP) {
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("AP mode supports WEP security only."));
+        }
+
+        // Initialize WiFi in AP mode.
+        if (nina_start_ap(ssid, security, key, channel) != 0) {
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("failed to start in AP mode"));
+        }
+    }
+    return mp_const_none;
+}
+
+static mp_obj_t py_nina_disconnect(mp_obj_t self_in) {
+    nina_disconnect();
+    return mp_const_none;
+}
+
+static mp_obj_t py_nina_isconnected(mp_obj_t self_in) {
+    return mp_obj_new_bool(nina_isconnected());
+}
+
+static mp_obj_t py_nina_connected_sta(mp_obj_t self_in) {
+    uint32_t sta_ip = 0;
+    mp_obj_t sta_list = mp_obj_new_list(0, NULL);
+
+    if (nina_connected_sta(&sta_ip) == 0) {
+        mp_obj_list_append(sta_list, netutils_format_inet_addr(
+            (uint8_t *)&sta_ip, 0, NETUTILS_BIG));
+    }
+
+    return sta_list;
+}
+
+static mp_obj_t py_nina_wait_for_sta(mp_obj_t self_in, mp_obj_t timeout_in) {
+    uint32_t sta_ip;
+    // get timeout
+    mp_obj_t sta_list = mp_obj_new_list(0, NULL);
+    mp_uint_t timeout = mp_obj_get_int(timeout_in);
+
+    if (nina_wait_for_sta(&sta_ip, timeout) == 0) {
+        mp_obj_list_append(sta_list, netutils_format_inet_addr(
+            (uint8_t *)&sta_ip, 0, NETUTILS_BIG));
+    }
+
+    return sta_list;
+}
+
+static mp_obj_t py_nina_ifconfig(size_t n_args, const mp_obj_t *args) {
+    nina_ifconfig_t ifconfig;
+    if (n_args == 1) {
+        // get ifconfig info
+        nina_ifconfig(&ifconfig, false);
+        mp_obj_t tuple[4] = {
+            netutils_format_ipv4_addr(ifconfig.ip_addr, NETUTILS_BIG),
+            netutils_format_ipv4_addr(ifconfig.subnet_addr, NETUTILS_BIG),
+            netutils_format_ipv4_addr(ifconfig.gateway_addr, NETUTILS_BIG),
+            netutils_format_ipv4_addr(ifconfig.dns_addr, NETUTILS_BIG),
+        };
+        return mp_obj_new_tuple(4, tuple);
+    } else {
+        // set ifconfig info
+        mp_obj_t *items;
+        mp_obj_get_array_fixed_n(args[1], 4, &items);
+        netutils_parse_ipv4_addr(items[0], ifconfig.ip_addr, NETUTILS_BIG);
+        netutils_parse_ipv4_addr(items[1], ifconfig.subnet_addr, NETUTILS_BIG);
+        netutils_parse_ipv4_addr(items[2], ifconfig.gateway_addr, NETUTILS_BIG);
+        netutils_parse_ipv4_addr(items[3], ifconfig.dns_addr, NETUTILS_BIG);
+        nina_ifconfig(&ifconfig, true);
+        return mp_const_none;
+    }
+}
+
+static mp_obj_t py_nina_netinfo(mp_obj_t self_in) {
```

Comment:
To match other network drivers' API, the settings returned from this function should instead be accessed via `.config('mac')` etc.

See https://docs.micropython.org/en/latest/library/network.html#network.AbstractNIC.status and https://docs.micropython.org/en/latest/library/network.html#network.AbstractNIC.config